FIX: Introduced new keyword 'density' in the hist function

moboehle opened the pull request:
Introduced new keyword 'density' in the hist function to be consistent with numpy.

This pull request closes issue #7364. Contrary to what was suggested in the comments, no error is raised in the case that both keywords density and normed are used. If neither are defined, default is 'False'.

QuLogic commented:
@tacaswell I thought we had a PR for this; did it not get merged?

phobson commented:
@QuLogic  see #7856

moboehle commented:
Sorry, I didn't see that there was a PR already, since it did not reference the issue.
The solutions to introducing 'density' as a new keyword are almost identical. However, as @tacaswell pointed out, we can pass the keyword density to np.histogram and delete the additional normalization in the hist function, in order to simplify the code. This is the main difference between the two fixes. Furthermore, in the current version of this PR, no error is raised if 'normed' and 'density' are both used but not contradicting. I think as long as there is no ambiguity in the input, no error should be raised, as it would cause an unnecessary inconvenience to the user.

phobson commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6220,7 +6225,7 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
                 slc = slice(None, None, -1)
 
-            if normed:
+            if density:
```
Looks like this path is never tested. Could you add a test that uses the `density` option?

phobson commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6132,6 +6135,10 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
         if histtype == 'barstacked' and not stacked:
             stacked = True
 
+        if density is not None and normed is not None and normed != density:
```
Add a test to check this

phobson commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6132,6 +6135,10 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
             stacked = True
 
+        if density is not None and normed is not None and normed != density:
+            raise ValueError("kwargs density and normed cannot have different\
```
Minor style comment. Can you do:

```python
raise ValueError("kwargs density and normed cannot have different "
                 "values. Use density only, since normed will be deprecated.")

tacaswell commented:
Thanks for your work on this!  

I definitely think it should raise if both kwargs are passed.  If the uses passes in both, then they must be passed as identical values (so we do not have to document and people do not have understand the precedence rules).   This can lead to very brittle code, will just move the frustration / debugging to a later time (when they two values get out of sync), and does not actually help improve the clarity.  

@moboehle Are you comfortable enough with git to merge the two pull requests together?

moboehle commented:
@phobson Thanks for your feedback, I added some tests now. 

@tacaswell You are right, allowing to set both 'normed' and 'density' could lead to odd behavior, I updated this now. I would be happy to merge the pull requests together, but I am not quite sure how to go about this.

dstansby commented on lib/matplotlib/pyplot.py:
```diff
@@ -2991,10 +2991,10 @@ def hexbin(x, y, C=None, gridsize=100, bins=None, xscale='linear',
-         bottom=None, histtype='bar', align='mid', orientation='vertical',
-         rwidth=None, log=False, color=None, label=None, stacked=False,
-         hold=None, data=None, **kwargs):
+def hist(x, bins=None, range=None, density=None, normed=None, weights=None,
```
`lib/matplotlib/pyplot.py` is an autogenerated file, so these changes will need to be reverted.

tacaswell commented:
#7856 has a test (that needs an image committed).

@moboehle I suggest you do the following:

```
git checkout bccn_hist # just to be safe
# add a remote
git remote add chelseatroy https://github.com/chelseatroy/matplotlib.git
# update your remotes
git remote update
git merge chelseatroy/histogram_compatibility_with_numpy_7364
```

You will probably have to clean up some conflicts, but then commit and push everything back to your branch.

You will need commit the test image results as well.

If you need help jump into https://gitter.im/matplotlib/matplotlib

QuLogic commented:
Ping @moboehle.

dstansby commented:
Hi @moboehle , thanks for the PR! Since there were two PRs I have merged them together with a few fixes and put them in #8993 - you will still be attributed to the work when it gets merged though.
